### PR TITLE
feat: download migrations using car files

### LINKED ISSUE #8851: Switch migration fetcher to CAR
go-ipfs 0.13 will add support for CAR responses (https://github.com/ipfs/go-ipfs/pull/8758)

When we have that running on the `ipfs.io` gateway, we should update [HttpFetcher](https://github.com/ipfs/go-ipfs/blob/83b3ec4a61f086f81e0f75c5ec9d508bf5385b16/repo/fsrepo/migrations/httpfetcher.go#L34) to fetch CARs instead of trusting gateway to send the correct bytes.

This will enable us to add more public gateway as a fallback in case DNS name gets censored etc without risking MITM.

The default gateways list should be ipfs.io, dweb.link, cloudflare-ipfs.com, cf-ipfs.com (+ make sure HTTP redirects are followed by the fetcher). We could also have the ultimate fallback which downloads https://github.com/ipfs/public-gateway-checker/raw/master/src/gateways.json and tries them too.

### PR DESCRIPTION
**Pull request by hacdias**
Close #8851.

This is an initial draft on replacing the migrations logic to use CAR files instead of retrieving the file directly from the gateway.

cc @lidel

### CODE REVIEW

**Comment by lidel** [comment]
cc @Jorropo so we avoid duplicated work :)

File: repo/fsrepo/migrations/httpfetcher.go
Comment (Jorropo) [inline repo/fsrepo/migrations/httpfetcher.go:153]: This is incorrect, basically you download a car file in order to hash the data locally however the root hash you are using is provided by the remote peer.
An attacker could give you a malicious root that include malware and you would be oblivious to it.
(it's fine because the attacker is the gateway which default to us, but this change would basically do nothing to the trust model, we want to stop trusting the gateway :slightly_smiling_face:)

You need to use the `path.Join(f.distPath, filePath)` above and perform the unixfs path walk manually, I can do that with feather if you want (I think this is gonna be easier.)

**Review by Jorropo — CHANGES_REQUESTED** [review]

File: repo/fsrepo/migrations/httpfetcher.go
Comment (Jorropo) [reply repo/fsrepo/migrations/httpfetcher.go:153]: > You need to use the `path.Join(f.distPath, filePath)` above and perform the unixfs path walk manually, I can do that with feather if you want (I think this is gonna be easier.)

Actually it might be possible to reuse handler code `ipfs get` with the in memory datastore.
Which wouldn't be that hard to do (it would be extremely ugly tho).

Comment (hacdias) [reply repo/fsrepo/migrations/httpfetcher.go:153]: @Jorropo if you're willing to, feel free to make any changes to this PR. I completely forgot about it!

**Comment by BigLep** [comment]
@hacdias @Jorropo : what are the next steps here?

**Comment by Jorropo** [comment]
The current `Roots[0]` is really wrong and essentially make this a security no-op.
I think the best solution is to wait until we land the required light clients in boxo. IPIP 402 would also make this work much easier (because currently we need to write individual block requests for path resolution and can only switch to car once the whole file has been reached in the DAG).

**Comment by hacdias** [comment]
I'm closing this now until we come up with a better solution.

**Comment by hacdias** [comment]
I also want to note that we also need verified (DNSLink) IPNS resolution as `filePath` might be IPNS and not an IPFS path. @Jorropo do you have a ETA for the light client to land in Boxo? I don't see any PR over there.

**Comment by hacdias** [comment]
@Jorropo okay, I just took a second look at this and indeed it'll be much easier once we have widespread IPIP-0402. There are two main problems here:

a. We need to be able to resolve `distPath`. If it is a `/ipfs` path, we're good to go. If it's a `/ipns` path we have to:
	1. Check if it's regular IPNS. If so, we can fetch record from gateway via `?format=ipns-record`, validate and get the `/ipfs` path.
	2. If it's DNSLink (most likely), setup a NameSys resolver and resolve it via DNS.
b. With the `distPath` being an `/ipfs` path, we can now request a CAR for `{distPath}{filePath}?dag-scope=entity` and verify if the CAR and path resolves correctly. Use that data for the migration file.

**Comment by hacdias** [comment]
Let's continue any discussion in https://github.com/ipfs/kubo/issues/9159.

**Pull request closed without merging**
